janus.c

```diff
@@ -203,7 +206,14 @@ void janus_handle_signal(int signum)
 			JANUS_PRINT("Ok, leaving immediately...\n");
 			break;
 	}
+
+	/* Wake the main loop */
+	janus_mutex_lock(&stop_mutex);
```

Comment:
It's generally not a good idea to lock a mutex from a signal handler. Some pthread_mutex_lock man pages mention this. Also you can see that there's a limited list of "async-signal safe functions" in the `man 7 signal` man page.

It's not hard to understand why: the same thread may have already taken the mutex, when the signal interrupts it and jumps it to the signal handler, where if it tries to take the mutex again and sleeps waiting for it to be available, it will deadlock.

One thing you *can* do here is write a single byte to a pipe. Or just let the while/sleep loop be, maybe making the sleep longer since it will be interrupted by a signal anyway.

(this comment moved from the commit to the pull request diff)